README

Code and documentation are copyright 2006-2008 Henning Norén

Parts of pdfcrack.c and md5.c is derived/copied/inspired from 
xpdf/poppler and are copyright 1995-2006 Glyph & Cog, LLC.

The PDF data structures, operators, and specification are
copyright 1985-2006 Adobe Systems Inc.


Project page: http://sourceforge.net/projects/pdfcrack/


pdfcrack is a simple tool for recovering passwords from pdf-documents.
It should be able to handle all pdfs that uses the standard security handler
but the pdf-parsing routines are a bit of a quick hack so you might stumble
across some pdfs where the parser needs to be fixed to handle.

Type 'make' (or 'gmake' if you have BSD-make as default) to build the program.
You will need to have GNU Make and a recent version of GCC installed but there
are no external dependencies on libraries.
You will have to add the -march-switch in the CFLAGS-option in Makefile
for best optimization on your platform. Look into the GCC-manual 
(http://gcc.gnu.org/onlinedocs/) if you are unsure.

The program is distributed under GPL version 2 (or later).

Features available in this release (check TODO for features that might come):
* Both owner- and user-passwords with the Standard Security Handler, rev 2 & 3.
* Search by wordlist
* Search by bruteforcing with specific charset
* Optimized search for owner-password when user-password is known (or empty)
* Extremely simple permutations of passwords (makes first letter uppercase)

- currently only useful for bruteforcing with charsets:
* Auto-save when interrupted (Ctrl-C or send SIGINT to the process)
* Loading saved state

- currently only for bruteforcing with charsets:
* Minimum length of password to start at
* Maximum length of password to try


Sort your wordlist by length for best performance and consider that almost
all passwords in PDFs are in iso latin 1 so use the correct character encoding
in your terminal and/or wordlist when using special characters.

This tool can not decrypt a Password Protected PDF. 
Look up the pdftk toolkit which can do that, when you know the password.


changelog
2020-04-24  Henning Norén  <henning.noren@gmail.com> - 0.19
* Fix MD5-calculation for very short keylengths (impossible in PDF-files)
* Do not allow longer RC5 than 128 bits due to undefined behavior in specs
  Thanks to p1ay8y3ar for reporting these issues
* Fix parsing issue for Encrypt object (thanks to Stefano Carbone for report)
* Fix a few signedness issues
* Bump version to 0.19
2019-08-19  Henning Norén  <henning.noren@gmail.com> - 0.18
* Update copyright years for all changed files
* Relax the search for encrypt object a bit more (thanks to Oleg for report)
* Tweak benchmark to make better comparison between md5_50 slow and fast
* Multiple inputfiles (-l) is not supported, so warn if used	
* Remove a bunch of unecessary if(ptr) before free(ptr)
* Bump version to 0.18
2019-04-17  Henning Norén  <henning.noren@gmail.com> - 0.17
* Ignore comments after obj in pdfparser. Thanks to Hima for report
* Make loading save-state files a bit more robust
  Thanks to Murmus CTF for showing these problems
* Fix incorrect fallthrough in pdfparser
* Bump version to 0.17
2017-04-19  Henning Norén  <henning.noren@gmail.com> - 0.16
* Remove incorrectly used pure-attribute (thanks to Carl Shapiro for report)
* Fix a possible overflow by one in the parser
* Fix pdfcrack starting despite being unable to extract encryption info
* Fix indention in man-page. Thanks to Jakub Wilk for patch
* Bump version to 0.16
2015-09-25  Henning Norén  <henning.noren@gmail.com> - 0.15
* Fix a few crashes found by "american fuzzy lop"
* Make the benchmark a bit more robust with volatile
* Fix a couple of undefined behaviours with left shift on int-type
* Fix GNU Make-isms in Makefile, thanks to Leonardo Taccari
* Fix various clang and cppchecker issues and warnings
* Bump version to 0.15
2014-08-31  Henning Norén  <henning.noren@gmail.com> - 0.14
* Add man-page
* s/CFLAGS=/CFLAGS +=/ in Makefile
* s/gcc/($CC)/ in Makefile
* Add support for wordlists and pdfs > 2GB (thanks to Gary Dewey)
* Bump version to 0.14
2014-02-10  Henning Norén  <henning.noren@gmail.com> - 0.13
* Skip strip and add debug symbols in Makefile
* Bail out if encountering rev 5+ (Not supported)
* Fix parsing of certain rev 4+ documents
* Add initial support for rev 5 (SHA256)
* Fix rev 3 owner password issues when other than 128 bit keylengths
* Update copyright dates in header of files
* Fix a gcc warning
* Fix filemode when opening a file on unix to make it more portable
  Thanks to Shinobu Maehara for the notice
* Fix a pdfparser logic bug, thanks to Shinobu Maehara for reporting
* bump version to 0.13
2014-02-01  Henning Norén  <henning.noren@gmail.com> - 0.12
* Print warning for odd key lengths
* Add support for handling Rev 3 with other than 128 bit keylengths
* Try parsing Encrypt object even when ID object is not found
* Fix a few gcc-warnings
* bump version to 0.12
2008-08-03  Henning Norén  <henning.noren@gmail.com> - 0.11
* Print out the user password when finding the ownerpassword
* Be less strict when parsing magic number
* Fixed two null-pointer dereferences
* Actually start the search for trailer in last KB
* bump version to 0.11
2008-05-01  Henning Norén  <henning.noren@gmail.com> - 0.10
* Add support for different newline-styles in wordlists
* Fix three nasty issues when parsing binary strings
  A big thanks to Will McCullen who reported and helped debug them
* Change step 6 when computing the encryption key to only apply to rev 3+
* Add warnings when the parser got a O/U binary string != 32 bytes
* Add support for non-hex ID strings
* Change error messages when not finding the ID or Encrypt object in trailer
* Try with empty password when bruteforcing from 0 characters (#1767244)
* Mention pdftk in the README (#1767245)
* Start search for the trailer in the last KB 
* bump version to 0.10
2007-10-23  Henning Norén  <henning.noren@gmail.com> - 0.9
* flush the buffers in printProgress
* make use of %zu when printf-ing size_t
* bump version to 0.9
2006-10-16  Henning Norén  <henning.noren.402@student.lu.se> - 0.8
* Removed fweb and Wextra from CFlags in Makefile to support gcc 3.3.5
* Added snippet about needing GNU Make in README
* Support version >3 Standard security handlers
* Actually notice when EncryptMetadata is false and print out it's value
* Print a warning when someone sets minpw > maxpw
* Fix some warnings
* bump version to 0.8
2006-07-08  Henning Norén  <henning.noren.402@student.lu.se> - 0.7
* Fix a problem with initializations of passwords when using max-length < 32.
  Thanks to Antonis Tassis for pointing it out.
* Fix some inconsistency when using --version-argument
* bump version to 0.7
2006-07-06  Henning Norén  <henning.noren.402@student.lu.se> - 0.6
* Actually save the permutation-setting when saving
* bump version to 0.6
2006-07-06  Henning Norén  <henning.noren.402@student.lu.se> - 0.5
* Make loading and saving states much more robust
* Block all signals when terminating from a SIGINT
* Add support for turning on permutations that is now off by default
* Various small tweaks, cleanups and bugfixes
* Optimizations of md5 and rc4
* bump version to 0.5
2006-06-18  Henning Norén  <henning.noren.402@student.lu.se> - 0.4
* Switch from using processor-time to using real time when reporting progress
* Add support for some simple benchmarking
* Add a check for the correct security-handler version to avoid a segfault
* Don't try the same password twice when permutating
* Fix a bug that made pdfcrack unable to load the upw from saved state file
* Fix a bug that ignored the saved max password when loading a saved state
* Various small fixes and changes
* bump version to 0.4
2006-05-28  Henning Norén  <henning.noren.402@student.lu.se> - 0.3
* Move CREDITS into README and add reference to project-page at sourceforge
* Various cleanups and comments
* Optimize md5 round 1 according to Colin Plumb's implementation
* bump version to 0.3
2006-05-26  Henning Norén  <henning.noren.402@student.lu.se> - 0.2

main.c
/**
 * Copyright (C) 2006-2020 Henning Norén
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, 
 * USA.
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <signal.h>
#include <getopt.h>
#include <string.h>
#include <unistd.h>
#include "pdfparser.h"
#include "pdfcrack.h"
#include "benchmark.h"

#ifdef __WIN32__
  #include <windows.h>
  #define	SIGALRM	14	/* alarm clock */
#endif 


#define PRINTERVAL 20 /** Print Progress Interval (seconds) */
#define CRASHFILE "savedstate.sav"
#define VERSION_MAJOR 0
#define VERSION_MINOR 19

#define _FILE_OFFSET_BITS 64

/** alarmInterrupt is used to print out the progress at specific intervals */
static void
alarmInterrupt() {
  if(!printProgress())
    alarm(PRINTERVAL);
}

/** autoSave is used to save down the current state when interrupted */
__attribute__ ((noreturn)) static void
autoSave(int sig) {
  FILE *file;

  if(sig)
    fprintf(stderr,"Caught signal %d!\nTrying to save state...\n",sig);
  if((file = fopen(CRASHFILE, "w")) == 0) {
    fprintf(stderr,"Could not open %s for writing\n", CRASHFILE);
  }
  else{
    saveState(file);
    fclose(file);
    fprintf(stderr,"Successfully saved state to %s!\n",CRASHFILE);
  }
  exit(sig);
}

/** print some help for the user */
static void
printHelp(char *progname) {
  printf("Usage: %s -f filename [OPTIONS]\n"
	 "OPTIONS:\n"
	 "-b, --bench\t\tperform benchmark and exit\n"
	 "-c, --charset=STRING\tUse the characters in STRING as charset\n"
	 "-w, --wordlist=FILE\tUse FILE as source of passwords to try\n"
	 "-n, --minpw=INTEGER\tSkip trying passwords shorter than this\n"
	 "-m, --maxpw=INTEGER\tStop when reaching this passwordlength\n"
	 "-l, --loadState=FILE\tContinue from the state saved in FILENAME\n"
	 "-o, --owner\t\tWork with the ownerpassword\n"
	 "-u, --user\t\tWork with the userpassword (default)\n"
	 "-p, --password=STRING\tGive userpassword to speed up breaking\n"
	 "\t\t\townerpassword (implies -o)\n"
	 "-q, --quiet\t\tRun quietly\n"
	 "-s, --permutate\t\tTry permutating the passwords (currently only\n"
	 "\t\t\tsupports switching first character to uppercase)\n"
	 "-v, --version\t\tPrint version and exit\n",
	 progname);
}

int
main(int argc, char** argv) {
  int ret = 0, minpw = 0, maxpw = 32;

  #ifndef __WIN32__
    struct sigaction act1, act2;
  #endif
  FILE *file = NULL, *wordlist = NULL;
  bool recovery = false, quiet = false, 
    work_with_user = true, permutation = false;
  uint8_t *userpassword = NULL;
  char *charset = NULL, *inputfile = NULL, *wordlistfile = NULL;
  EncData *e = NULL;

  /** Parse arguments */
  while(true) {
    int c, option_index;
    static struct option long_options[] = {
      {"bench",    no_argument      , 0, 'b'},
      {"charset",  required_argument, 0, 'c'},
      {"file",     required_argument, 0, 'f'},
      {"loadState",required_argument, 0, 'l'},
      {"maxpw",    required_argument, 0, 'm'},
      {"minpw",    required_argument, 0, 'n'},
      {"owner",    no_argument      , 0, 'o'},
      {"password", required_argument, 0, 'p'},
      {"quiet",    required_argument, 0, 'q'},     
      {"permutate",no_argument,       0, 's'},
      {"user",     no_argument      , 0, 'u'},
      {"wordlist", required_argument, 0, 'w'},
      {"version",  no_argument      , 0, 'v'},     
      {0, 0, 0, 0}};
    /* getopt_long stores the option index here. */
    option_index = 0;
     
    c = getopt_long(argc, argv, "bc:f:l:m:n:op:qsuw:v",
			long_options, &option_index);
    
    /* Detect the end of the options. */
    if(c == -1)
      break;
    
    switch(c) {
    case 'b':
      runBenchmark();
      return 0;
    case 'c':
      if(charset)
	fprintf(stderr, "Charset already set\n");
      else
	charset = strdup(optarg);
      break;
    case 'f':
      if(!recovery)
	inputfile = strdup(optarg);
      break;

    case 'l':
      if(inputfile) {
	fprintf(stderr, "Inputfile already set\n");
      }
      else {
	inputfile = strdup(optarg);
	recovery = true;
      }
      break;
	
    case 'm':
      maxpw = atoi(optarg);
      break;
     
    case 'n':
      minpw = atoi(optarg);
      break;

    case 'o':
      work_with_user = false;
      break;

    case 'p':
      if(userpassword)
	fprintf(stderr, "User password already set\n");
      else {
	userpassword = (uint8_t*)strdup(optarg);
	work_with_user = false;
      }
      break;

    case 'q':
      quiet = true;
      break;

    case 'u':
      if(!work_with_user) {
	fprintf(stderr, "Cannot work with both user- and owner-password\n");
	exit(1);
      }
      work_with_user = true;
      break;

    case 's':
      permutation = true;
      break;

    case 'w':
      if(wordlistfile)
	fprintf(stderr, "Wordlist already set\n");
      else
	wordlistfile = strdup(optarg);
      break;

    case 'v':
      printf("pdfcrack version %d.%d\n", VERSION_MAJOR, VERSION_MINOR);
      return 0;

    default:
      printHelp(argv[0]);
      ret = 1;
      goto out2;
    }
  }

  /** If there are any unhandled arguments and the filename and/or wordlist is
      not set: try to match the first as the filename and the second as 
      wordlist.
  */
  {
    int i = optind;
    if(i > 0) {
      if(i < argc && !inputfile)
	inputfile = strdup(argv[i++]);
      if(i < argc && !wordlistfile)
	wordlistfile = strdup(argv[i++]);
      if(i < argc)
	while(i<argc) {
	  fprintf(stderr,"Non-option argument %s\n", argv[i++]);
	}
    }
  }

  if(!inputfile || minpw < 0 || maxpw < 0) {
    printHelp(argv[0]);
    ret = 1;
    goto out2;
  }

  if((file = fopen(inputfile, "rb")) == 0) {
    fprintf(stderr,"Error: file %s not found\n", inputfile);
    ret = 2;
    goto out2;
  }

  e = calloc(1,sizeof(EncData));

  if(recovery) {
    if(wordlistfile) {
      free(wordlistfile);
      wordlistfile = NULL;
    }
    if(!loadState(file, e, &wordlistfile, &work_with_user)) {
      fprintf(stderr, "Error: Not a savefile or savefile is damaged\n");
      ret = 3;
      goto out1;
    }

    if(!quiet)
      printf("Loaded state for %s\n", inputfile);
  }
  else { /** !recovery */
    if(!openPDF(file,e)) {
      fprintf(stderr, "Error: Not a valid PDF\n");
      ret = 3;
      goto out1;
    }

    ret = getEncryptedInfo(file, e);
    if(ret) {
      if(ret == EENCNF) 
	fprintf(stderr, "Error: Could not extract encryption information\n");
      else if(ret == ETRANF || ret == ETRENF || ret == ETRINF)
	fprintf(stderr, "Error: Encryption not detected (is the document password protected?)\n");
      ret = 4;
      goto out1;
    }
    else if(e->revision < 2 || (strcmp(e->s_handler,"Standard") != 0 || e->revision > 5)) {
      fprintf(stderr, "The specific version is not supported (%s - %d)\n", e->s_handler, e->revision);
      ret = 5;
      goto out1;
      }
  }

  if(fclose(file)) {
    fprintf(stderr, "Error: closing file %s\n", inputfile);
  }

  if(minpw > maxpw) {
    fprintf(stderr, "Warning: minimum pw-length bigger than max\n");
  }

  if(wordlistfile) {
    if((wordlist = fopen(wordlistfile, "r")) == 0) {
      fprintf(stderr,"Error: file %s not found\n", wordlistfile);
      ret = 6;
      goto out2;
    }
  }

#ifndef __WIN32__
  act2.sa_handler = autoSave;
  sigfillset(&act2.sa_mask);
  act2.sa_flags = 0;
  sigaction(SIGINT, &act2, 0);
#else
    signal(SIGINT, autoSave);
#endif

  if(!quiet) {
    printEncData(e);
#ifndef __WIN32__
    act1.sa_handler = alarmInterrupt;
    sigemptyset(&act1.sa_mask);
    act1.sa_flags = 0;
    sigaction(SIGALRM, &act1, 0);
#else
    signal(SIGALRM, alarmInterrupt);
    
#endif
  alarm(PRINTERVAL);
  }


  /** Try to initialize the cracking-engine */
  if(!initPDFCrack(e, userpassword, work_with_user, wordlistfile,
		   wordlistfile?Wordlist:Generative, wordlist, charset, 
		   (unsigned int)minpw, (unsigned int)maxpw, permutation)) {
    cleanPDFCrack();
    fprintf(stderr, "Wrong userpassword, '%s'\n", userpassword);
    ret = 7;
    goto out2;
  }

  /** Do the actual crunching */
  runCrack();


  /** cleanup */
  cleanPDFCrack();
  freeEncData(e);

  if(wordlistfile) {
    fclose(wordlist);
    free(wordlistfile);
  }
  free(inputfile);
  free(charset);
  free(userpassword);

  return 0;

 out1:
  if(fclose(file))
    fprintf(stderr, "Error: closing file %s\n", inputfile);
 out2:
  freeEncData(e);
  free(inputfile);
  free(charset);
  free(userpassword);

  exit(ret);
}
